scheds/rust/scx_layered/src/metrics.rs

```diff
@@ -0,0 +1,197 @@
+use prometheus_client::metrics::gauge::Gauge;
+use prometheus_client::registry::Registry;
+
+#[derive(Default)]
+pub struct OpenMetricsStats {
+    pub registry: Registry,
+    pub total: Gauge<i64, AtomicI64>,
+    pub local: Gauge<f64, AtomicU64>,
+    pub open_idle: Gauge<f64, AtomicU64>,
+    pub affn_viol: Gauge<f64, AtomicU64>,
+    pub excl_idle: Gauge<f64, AtomicU64>,
+    pub excl_wakeup: Gauge<f64, AtomicU64>,
+    pub proc_ms: Gauge<i64, AtomicI64>,
+    pub busy: Gauge<f64, AtomicU64>,
+    pub util: Gauge<f64, AtomicU64>,
+    pub load: Gauge<f64, AtomicU64>,
+    pub l_util: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_util_frac: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_load: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_load_frac: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_tasks: Family<Vec<(String, String)>, Gauge<i64, AtomicI64>>,
+    pub l_total: Family<Vec<(String, String)>, Gauge<i64, AtomicI64>>,
+    pub l_sel_local: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_enq_wakeup: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_enq_expire: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_enq_last: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_enq_reenq: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_min_exec: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_min_exec_us: Family<Vec<(String, String)>, Gauge<i64, AtomicI64>>,
+    pub l_open_idle: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_preempt: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_preempt_first: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_preempt_idle: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_preempt_fail: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_affn_viol: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_keep: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_keep_fail_max_exec: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_keep_fail_busy: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_excl_collision: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_excl_preempt: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_kick: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_yield: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_yield_ignore: Family<Vec<(String, String)>, Gauge<i64, AtomicI64>>,
+    pub l_migration: Family<Vec<(String, String)>, Gauge<f64, AtomicU64>>,
+    pub l_cur_nr_cpus: Family<Vec<(String, String)>, Gauge<i64, AtomicI64>>,
+    pub l_min_nr_cpus: Family<Vec<(String, String)>, Gauge<i64, AtomicI64>>,
+    pub l_max_nr_cpus: Family<Vec<(String, String)>, Gauge<i64, AtomicI64>>,
+}
+
+impl OpenMetricsStats {
```

Comment:
I almost wish this had a `report` method that was most of the logic from the `Sched::report` method, but that's too much refactoring.